File: cellviz/src/cellviz/plots/ica.py
from __future__ import annotations

from typing import Any, Literal

import plotly.graph_objects as go

from ..compute import compute_ica_dvq_per_cycle
from ..io import CellData, CyclerName, load_many
from ..style import cell_color, cycle_fade_color


Direction = Literal["discharge", "charge"]


def _ica_points(cell: CellData, cycles: list[int] | None, direction: Direction) -> list[tuple[int, list[float], list[float]]]:
    computed = compute_ica_dvq_per_cycle(cell.curves, direction=direction)
    out: list[tuple[int, list[float], list[float]]] = []
    allowed = set(cycles) if cycles else None
    for cycle, payload in sorted(computed.items()):
        if allowed is not None and cycle not in allowed:
            continue
        ica = payload.get("ica", {})
        v = list(ica.get("v", []))
        dqdv = list(ica.get("dqdv", []))
        if len(v) >= 2 and len(dqdv) >= 2:
            out.append((cycle, v, dqdv))
    return out


def _voltage_range(per_cell: list[list[tuple[int, list[float], list[float]]]]) -> tuple[float, float] | None:
    all_v: list[float] = []
    for rows in per_cell:
        for _, v, _ in rows:
            all_v.extend(v)
    if not all_v:
        return None
    return max(all_v), min(all_v)


def _build_traces_3d(cells: list[CellData], per_cell: list[list[tuple[int, list[float], list[float]]]]) -> list[go.Scatter3d]:
    traces: list[go.Scatter3d] = []
    for i, (cell, rows) in enumerate(zip(cells, per_cell)):
        base = cell.color or cell_color(cell.cell_id or cell.label or f"dataset-{i+1}", i)
        total = len(rows)
        for j, (cycle, v, dqdv) in enumerate(rows):
            traces.append(
                go.Scatter3d(
                    x=v,
                    y=[cycle] * len(v),
                    z=dqdv,
                    mode="lines",
                    name=f"{cell.label} - Cycle {cycle}",
                    legendgroup=cell.cell_id or cell.label or f"group-{i}",
                    showlegend=j == total - 1,
                    line={"color": cycle_fade_color(base, 1 if total <= 1 else j / (total - 1)), "width": 3},
                    hovertemplate=f"{cell.label}<br>Voltage: %{{x:.2f}} V<br>Cycle: %{{y}}<br>IC: %{{z:.2f}} Ah/V<extra></extra>",
                )
            )
    return traces


def _find_peak(values: list[float]) -> tuple[int, float]:
    idx = max(range(len(values)), key=lambda i: values[i])
    return idx, values[idx]


def _build_traces_2d(cells: list[CellData], per_cell: list[list[tuple[int, list[float], list[float]]]]) -> list[go.Scatter]:
    traces: list[go.Scatter] = []
    for i, (cell, rows) in enumerate(zip(cells, per_cell)):
        if not rows:
            continue
        base = cell.color or cell_color(cell.cell_id or cell.label or f"dataset-{i+1}", i)
        baseline = rows[1] if len(rows) > 1 else None
        selected = rows[-1]
        if baseline:
            traces.append(
                go.Scatter(
                    x=baseline[1],
                    y=baseline[2],
                    mode="lines",
                    name=f"{cell.label} Cycle {baseline[0]}",
                    line={"color": "rgba(31,119,180,0.4)" if base == "#1f77b4" else base, "width": 1.5},
                    hovertemplate=f"{cell.label} Cycle {baseline[0]}<br>Voltage: %{{x:.2f}} V<br>IC: %{{y:.2f}} Ah/V<extra></extra>",
                )
            )
        peak_idx, peak_val = _find_peak(selected[2])
        peak_v = selected[1][peak_idx]
        traces.append(
            go.Scatter(
                x=selected[1],
                y=selected[2],
                mode="lines",
                name=cell.label,
                line={"color": base, "width": 2},
                hovertemplate=f"{cell.label}<br>Voltage: %{{x:.2f}} V<br>IC: %{{y:.2f}} Ah/V<br>Peak: {peak_v:.2f} V, {peak_val:.2f} Ah/V<extra></extra>",
            )
        )
        traces.append(
            go.Scatter(
                x=[peak_v],
                y=[peak_val],
                mode="markers",
                name=f"{cell.label} peak",
                marker={"size": 10, "color": base, "symbol": "diamond", "line": {"width": 1, "color": "#fff"}},
                hovertemplate=f"{cell.label} peak<br>Voltage: %{{x:.2f}} V<br>IC: %{{y:.2f}} Ah/V<extra></extra>",
            )
        )
    return traces


def _scene(v_range: tuple[float, float] | None) -> dict[str, Any]:
    return {
        "xaxis": {"title": {"text": "Voltage (V)"}, "range": list(v_range) if v_range else None},
        "yaxis": {"title": {"text": "Cycle"}},
        "zaxis": {"title": {"text": "dQ/dV (Ah/V)"}},
        "camera": {"eye": {"x": 1.25, "y": 1.25, "z": 1.25}},
    }


def ica_plot(
    data: Any,
    *,
    cycles: list[int] | None = None,
    mode: str = "3d",
    labels: list[str] | None = None,
    direction: Direction = "discharge",
    cycler: CyclerName = "auto",
) -> go.Figure:
    cells = load_many(data, labels=labels, cycler=cycler)
    per_cell = [_ica_points(cell, cycles, direction) for cell in cells]
    if all(len(rows) == 0 for rows in per_cell):
        raise ValueError(
            "No valid ICA cycles found. Ensure input contains charge/discharge records "
            "with non-zero capacity change and voltage span."
        )

    if mode == "2d":
        fig = go.Figure(data=_build_traces_2d(cells, per_cell))
        fig.update_layout(
            xaxis={"title": {"text": "Voltage (V)"}},
            yaxis={"title": {"text": "Incremental Capacity (Ah/V)"}},
            showlegend=True,
        )
        return fig

    fig = go.Figure(data=_build_traces_3d(cells, per_cell))
    fig.update_layout(scene=_scene(_voltage_range(per_cell)), showlegend=True)
    return fig

File: frontend/src/cellviz-lib/plotting/ica.ts
import { cellColor, cycleFadeColor } from '../style';
import type { BuildIcaOpts, Dataset, Ica2DViewMode, IcaFigure } from '../types';

function hexToRgba(hex: string, opacity: number): string {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  const a = Math.max(0.3, Math.min(1, opacity));
  return `rgba(${r},${g},${b},${a})`;
}

function findPeak(arr: number[]): { index: number; value: number } {
  let index = 0;
  let value = arr[0] ?? 0;
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] > value) {
      value = arr[i];
      index = i;
    }
  }
  return { index, value };
}

function baseColor(d: Dataset, idx: number): string {
  return d.color ?? cellColor(d.id, idx);
}

function voltageExtent(datasets: Dataset[]): [number, number] | undefined {
  let min = Number.POSITIVE_INFINITY;
  let max = Number.NEGATIVE_INFINITY;
  for (const dataset of datasets) {
    for (const cycle of dataset.cycles) {
      for (const v of cycle.x) {
        if (v < min) min = v;
        if (v > max) max = v;
      }
    }
  }
  if (!Number.isFinite(min) || !Number.isFinite(max)) return undefined;
  return [max, min];
}

function smallFont(): { size: number; family: 'Inter' } {
  return { size: 10, family: 'Inter' };
}

function build3D(datasets: Dataset[]): IcaFigure {
  const traces: Plotly.Data[] = [];
  datasets.forEach((dataset, i) => {
    const color = baseColor(dataset, i);
    const total = dataset.cycles.length;
    dataset.cycles.forEach((cycle, idx) => {
      traces.push({
        x: cycle.x,
        y: new Array(cycle.x.length).fill(cycle.cycle),
        z: cycle.y,
        type: 'scatter3d',
        mode: 'lines',
        name: `${dataset.label} - Cycle ${cycle.cycle}`,
        line: { color: cycleFadeColor(color, total <= 1 ? 1 : idx / (total - 1)), width: 3 },
        legendgroup: dataset.id,
        showlegend: idx === total - 1,
        hovertemplate: `${dataset.label}<br>Voltage: %{x:.2f} V<br>Cycle: %{y}<br>IC: %{z:.2f} Ah/V<extra></extra>`,
      });
    });
  });

  const font = smallFont();
  return {
    data: traces,
    layout: {
      font,
      scene: {
        xaxis: { title: { text: 'Voltage (V)', font }, range: voltageExtent(datasets), gridcolor: '#e0e0e0', tickfont: font },
        yaxis: { title: { text: 'Cycle', font }, gridcolor: '#e0e0e0', tickfont: font },
        zaxis: { title: { text: 'dQ/dV (Ah/V)', font }, gridcolor: '#e0e0e0', tickfont: font },
        camera: { eye: { x: 1.25, y: 1.25, z: 1.25 } },
      },
      legend: { x: 0, y: -0.12, orientation: 'h', font },
      margin: { t: 36, r: 8, b: 8, l: 8 },
    },
  };
}

function build2D(datasets: Dataset[], opts: BuildIcaOpts): IcaFigure {
  const cycleIndex = opts.cycleIndex ?? 0;
  const viewMode: Ica2DViewMode = opts.viewMode ?? 'baseline';
  const baselineCycle = opts.baselineCycle;
  const baselineCycleIndex = opts.baselineCycleIndex ?? 1;
  const traces: Plotly.Data[] = [];

  if (viewMode === 'range') {
    const rangeTraces: Plotly.Data[] = [];
    const lines: Plotly.Data[] = [];
    const peaks: Plotly.Data[] = [];
    datasets.forEach((dataset, i) => {
      const color = baseColor(dataset, i);
      const xParts: (number | null)[] = [];
      const yParts: (number | null)[] = [];
      dataset.cycles.forEach((cycle, idx) => {
        if (idx > 0) {
          xParts.push(null);
          yParts.push(null);
        }
        xParts.push(...cycle.x);
        yParts.push(...cycle.y);
      });
      rangeTraces.push({
        x: xParts,
        y: yParts,
        type: 'scatter',
        mode: 'lines',
        name: `${dataset.label} range`,
        line: { color: hexToRgba(color, 0.25), width: 1.5 },
        hovertemplate: `${dataset.label}<br>Voltage: %{x:.2f} V<br>IC: %{y:.2f} Ah/V<extra></extra>`,
        showlegend: true,
        connectgaps: false,
      });

      const cycle = dataset.cycles[cycleIndex];
      if (!cycle) return;
      lines.push({
        x: cycle.x,
        y: cycle.y,
        type: 'scatter',
        mode: 'lines',
        name: `${dataset.label} Cycle ${cycle.cycle}`,
        line: { color, width: 2 },
        hovertemplate: `${dataset.label}<br>Voltage: %{x:.2f} V<br>IC: %{y:.2f} Ah/V<extra></extra>`,
      });

      const { index: peakIndex, value: peakValue } = findPeak(cycle.y);
      peaks.push({
        x: [cycle.x[peakIndex]],
        y: [peakValue],
        type: 'scatter',
        mode: 'markers',
        name: `${dataset.label} peak`,
        marker: { size: 10, color, symbol: 'diamond', line: { width: 1, color: '#fff' } },
        hovertemplate: `${dataset.label} peak<br>Voltage: %{x:.2f} V<br>IC: %{y:.2f} Ah/V<extra></extra>`,
      });
    });
    traces.push(...rangeTraces, ...lines, ...peaks);
  } else {
    const baselineLines: Plotly.Data[] = [];
    const lines: Plotly.Data[] = [];
    const peaks: Plotly.Data[] = [];
    datasets.forEach((dataset, i) => {
      const color = baseColor(dataset, i);
      const cycle = dataset.cycles[cycleIndex];
      if (!cycle) return;
      const baseline = baselineCycle != null
        ? dataset.cycles.find((c) => c.cycle === baselineCycle)
        : dataset.cycles[baselineCycleIndex];
      if (baseline) {
        baselineLines.push({
          x: baseline.x,
          y: baseline.y,
          type: 'scatter',
          mode: 'lines',
          name: `${dataset.label} Cycle ${baseline.cycle}`,
          line: { color: hexToRgba(color, 0.4), width: 1.5 },
          hovertemplate: `${dataset.label} Cycle ${baseline.cycle}<br>Voltage: %{x:.2f} V<br>IC: %{y:.2f} Ah/V<extra></extra>`,
        });
      }
      const { index: peakIndex, value: peakValue } = findPeak(cycle.y);
      const peakVoltage = cycle.x[peakIndex] ?? 0;
      lines.push({
        x: cycle.x,
        y: cycle.y,
        type: 'scatter',
        mode: 'lines',
        name: dataset.label,
        line: { color, width: 2 },
        hovertemplate: `${dataset.label}<br>Voltage: %{x:.2f} V<br>IC: %{y:.2f} Ah/V<br>Peak: ${peakVoltage.toFixed(2)} V, ${peakValue.toFixed(2)} Ah/V<extra></extra>`,
      });
      peaks.push({
        x: [peakVoltage],
        y: [peakValue],
        type: 'scatter',
        mode: 'markers',
        name: `${dataset.label} peak`,
        marker: { size: 10, color, symbol: 'diamond', line: { width: 1, color: '#fff' } },
        hovertemplate: `${dataset.label} peak<br>Voltage: %{x:.2f} V<br>IC: %{y:.2f} Ah/V<extra></extra>`,
      });
    });
    traces.push(...baselineLines, ...lines, ...peaks);
  }

  const font = smallFont();
  return {
    data: traces,
    layout: {
      font,
      xaxis: { title: { text: 'Voltage (V)', font }, tickfont: font, gridcolor: '#e0e0e0', showgrid: true },
      yaxis: { title: { text: 'Incremental Capacity (Ah/V)', font }, tickfont: font, gridcolor: '#e0e0e0', showgrid: true },
      legend: { x: 0.5, y: -0.12, xanchor: 'center', yanchor: 'top', orientation: 'h', font },
      margin: { t: 36, r: 8, b: 48, l: 40 },
      showlegend: true,
    },
  };
}

export function buildIcaFigure(datasets: Dataset[], opts: BuildIcaOpts = {}): IcaFigure {
  const mode = opts.mode ?? '3d';
  return mode === '2d' ? build2D(datasets, opts) : build3D(datasets);
}
